pkg/kube/krt/collection.go

```diff
@@ -226,11 +223,9 @@ func (h *manyCollection[I, O]) index(extract func(o O) []string) kclient.RawInde
 // onPrimaryInputEvent takes a list of I's that changed and reruns the handler over them.
 // This is called either when I directly changes, or if a secondary dependency changed. In this case, we compute which I's depended
 // on the secondary dependency, and call onPrimaryInputEvent with them
-func (h *manyCollection[I, O]) onPrimaryInputEvent(items []Event[I], lock bool) {
-	if lock {
-		h.recomputeMu.Lock()
-		defer h.recomputeMu.Unlock()
-	}
+func (h *manyCollection[I, O]) onPrimaryInputEvent(items []Event[I]) {
+	h.blockNewEvents.Lock()
+	defer h.blockNewEvents.Unlock()
```

Comment:
why do we need the lock if all these items are readonly

Reply:
I've updated the documentation in the code, but the idea is to make it so we can register handlers safely. When we add a have to send them `[initial state, any future updates]`. if an even is happening while we are sending the initial state, things get wonky. This ensures we can properly send these events. If you look at the k8s client-go informer code you will see the exact same mutex used (with the same name)

Reply:
my mistake it's a different name https://github.com/kubernetes/client-go/blob/37045084c2aa82927b0e5ffc752861430fd7e4ab/tools/cache/shared_informer.go#L389

Reply:
In k8s it is to protect event handler register. But here in istio, it seems to protect items. That's my confusion, items not need to protect when being distribute. If one element being changed while it is being distributed, it doesnot matter. This is the older one, and the new one will trigger another distribute

Reply:
Here it is also to protect event handler registration. In both cases it's used the same as far as I can tell - it blocks new events from coming in so that we cleanly register a handler?

Reply:
In istio  event handler registration is not protected with this lock, it is with mu

new events won't come in

Reply:
L587 is where the handler registration is protected with this lock?

Reply:
L160 is registering
```
	// Send out all the initial objects to the handler. We will then unlock the new events so it gets the future updates.
	return h.eventHandlers.Insert(f, h.Synced(), events, h.stop)
```

Reply:
I don't understand. that is line 600 not 160 and it's inside of the lock...

Reply:
.Insert method has a lock, i think we donot need blockNewEvents, will have a try

Reply:
Thanks for the commit, much appreciated. While the current set of tests doesn't catch a difference between this and your commit, I worked on a new test that does. It fails without the lock.

Let me show a timeline for why. (t1/t2=thread1 / thread 2, executing concurrently)

* t1: Add(a). Take `mu`, add the object, release `mu`.
* t2: Register. Take `mu`, get all the objects (`[a]`), send `[a]` to event handler. Note: the mu could be released before or after 'send [a]', it doesn't impact things; your PR moved it to release after the event handler.
* t1: continue Add(a) execution. Now we trigger send `[a]` to event handler.

Result: the event handlers gets `[a]` twice.

Here is an example failure locally when I remove the blockNewEvents lock:

```
=== RUN   TestConformance
=== RUN   TestConformance/informer
2024-11-27T15:58:57.509116Z     info    cluster "fake" kube client started
2024-11-27T15:58:57.511587Z     info    krt     Informer[*v1.ConfigMap] synced  owner=Informer[*v1.ConfigMap]
2024-11-27T15:58:57.511610Z     info    controllers     starting        controller=crd watcher
2024-11-27T15:58:57.837297Z     info    controllers     stopped controller=crd watcher
=== RUN   TestConformance/static_list
=== RUN   TestConformance/join
2024-11-27T15:58:58.161624Z     info    krt     Join[krt_test.Named] synced
=== RUN   TestConformance/manyCollection
2024-11-27T15:58:58.481589Z     info    krt     ManyCollection[string,krt_test.Named] synced    owner=ManyCollection[string,krt_test.Named]
    conformance_test.go:209: unexpected events: [add/a/12 add/a/13 add/a/14 add/a/15 add/a/16 add/a/19 add/a/17 add/a/18 add/a/11]
--- FAIL: TestConformance (1.84s)
    --- PASS: TestConformance/informer (0.87s)
    --- PASS: TestConformance/static_list (0.32s)
    --- PASS: TestConformance/join (0.32s)
    --- FAIL: TestConformance/manyCollection (0.32s)
FAIL
```

Here you can see it got duplicate events.

Reply:
Yes. Isn't this the expected result for eventual model, upstream k8s can also trigger twice with two events on `a`

Reply:
The Kubernetes informer cannot duplicate events. The eventually consistency doesn't mean you can get multiple events, it just means you may miss intermediate states (so if a pod goes from v1->v2->v3, you may just see v1->v3).

Reply:
If the client consumes fast enough, definitely it can handle  v1->v2->v3 

And if istio handles  v1->v2->v3, any caveats bring?

Reply:
I am just state the properties that Kubernetes guarantees. Like in the core, not even in client libraries. That it doesn't guarantee you see every update (though its super rare, and does not matter to Istio at all).

None of this has anything to do with this PR, so sorry if it added confusion. The main point I was trying to make is:
* Informers does **not** ever give duplicate events
* This PR does **not** ever give duplicate events
* The proposal (to drop the lock) **does** give duplicate events, which I consider a bug

Reply:
By duplicate event, assume you mean a handler can receive v1->v1->v2->v3. 

Is there a test can verify it? I cannot understand this possibility

Reply:
Yes I added the test https://github.com/istio/istio/pull/53994/commits/2ad9028412ae9cd010eea59abf851eff427f426c. It fails with the lock removed (inconsistently since it's a race). I showed a failure in this comment https://github.com/istio/istio/pull/53994#discussion_r1860927390

Reply:
Weird, i run go race test 30 times, all passed

Reply:
30 times is not often, took me 500